File: p34.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        // size one edge case
        if(nums.size() == 1) {
            if(nums[0] == target)
                return {0, 0};
            else
                return {-1, -1};
        }
        // find start point
        int start = -1;
        int n = nums.size();
        int left = 0;
        int right = n;
        int mid = right/2;
        while(left < right) {
            if(nums[mid] == target && (mid == 0 || nums[mid-1] < target)) {
                start = mid;
                break;
            } else if(left == mid || right == mid) {
                break;
            } else if(nums[mid] >= target) {
                right = mid;
                mid = (left + right)/2;
            } else {
                left = mid;
                mid = (left + right)/2;
            }
        }
        // find end point
        int end = -1;
        left = 0;
        right = n;
        mid = right/2;
        while(left < right) {
            if(nums[mid] == target && (mid == n-1 || nums[mid+1] > target)) {
                end = mid;
                break;
            } else if (left == mid || right == mid) {
                break;
            } else if(nums[mid] <= target) {
                left = mid;
                mid = (left + right)/2;
            } else {
                right = mid;
                mid = (left + right)/2;
            }
        }
        return {start, end};
    }
};

File: p34.py
class Solution:
    def binSearchLeft(self, nums: List[int], target: int, left: int, right: int) -> int:
        if left == right:
            return left if nums[left] == target else -1
        elif left+1 == right:
            if nums[left] == target:
                return left
            return right if nums[right] == target else -1
        mid = (left+right)//2
        if nums[mid] < target:
            left = mid
        else:
            right = mid
        return self.binSearchLeft(nums, target, left, right)
    
    def binSearchRight(self, nums: List[int], target: int, left: int, right: int) -> int:
        if left == right:
            return right if nums[right] == target else -1
        elif left+1 == right:
            if nums[right] == target:
                return right
            return left if nums[left] == target else -1
        mid = (left+right)//2
        if nums[mid] > target:
            right = mid
        else:
            left = mid
        return self.binSearchRight(nums, target, left, right)
        
    def searchRange(self, nums: List[int], target: int) -> List[int]:        
        solution = [-1,-1]
        if len(nums) == 0:
            return solution
        n = len(nums)
        solution[0] = self.binSearchLeft(nums, target, 0, n-1)
        solution[1] = self.binSearchRight(nums, target, 0, n-1)
        return solution